distributed/shuffle/tests/test_rechunk.py

```diff
@@ -0,0 +1,785 @@
+
+    local_shuffle_pool = ArrayRechunkTestPool()
+    shuffles = []
+    for i in range(n_workers):
+        shuffles.append(
+            local_shuffle_pool.new_shuffle(
+                name=workers[i],
+                worker_for_mapping=worker_for_mapping,
+                old=old,
+                new=new,
+                directory=tmpdir,
+                loop=loop_in_thread,
+            )
+        )
+    random.seed(42)
+    if barrier_first_worker:
+        barrier_worker = shuffles[0]
+    else:
+        barrier_worker = random.sample(shuffles, k=1)[0]
+
+    try:
+        for i, (idx, arr) in enumerate(old_chunks.items()):
+            s = shuffles[i % len(shuffles)]
+            await s.add_partition(arr, idx)
+
+        await barrier_worker.barrier()
+
+        total_bytes_sent = 0
+        total_bytes_recvd = 0
+        total_bytes_recvd_shuffle = 0
+        for s in shuffles:
+            metrics = s.heartbeat()
+            assert metrics["comm"]["total"] == metrics["comm"]["written"]
+            total_bytes_sent += metrics["comm"]["written"]
+            total_bytes_recvd += metrics["disk"]["total"]
+            total_bytes_recvd_shuffle += s.total_recvd
+
+        assert total_bytes_recvd_shuffle == total_bytes_sent
+
+        all_chunks = np.empty(tuple(len(dim) for dim in new), dtype="O")
+        for ix, worker in worker_for_mapping.items():
+            s = local_shuffle_pool.shuffles[worker]
+            all_chunks[ix] = await s.get_output_partition(ix)
+
+    finally:
+        await asyncio.gather(*[s.close() for s in shuffles])
+
+    old_cs = np.empty(tuple(len(dim) for dim in old), dtype="O")
+    for ix, arr in old_chunks.items():
+        old_cs[ix] = arr
+    np.testing.assert_array_equal(
+        concatenate3(old_cs.tolist()), concatenate3(all_chunks.tolist()), strict=True
+    )
+
+
+@pytest.mark.parametrize("rechunk_config", ["tasks", "p2p", None])
+@pytest.mark.parametrize("rechunk_keyword", ["tasks", "p2p", None])
+@gen_cluster(client=True, config={"optimization.fuse.active": False})
+async def test_rechunk_configuration(c, s, *ws, rechunk_config, rechunk_keyword):
+    """Try rechunking a random 1d matrix
+
+    See Also
+    --------
+    dask.array.tests.test_rechunk.test_rechunk_1d
+    """
+    a = np.random.uniform(0, 1, 30)
+    x = da.from_array(a, chunks=((10,) * 3,))
+    new = ((6,) * 5,)
+    with dask.config.set(rechunk=rechunk_config):
+        x2 = rechunk(x, chunks=new, rechunk=rechunk_keyword)
+    expected_method = rechunk_keyword if rechunk_keyword is not None else rechunk_config
+    if expected_method == "p2p":
+        assert all(key[0].startswith("rechunk-p2p") for key in x2.__dask_keys__())
+    else:
+        assert not any(key[0].startswith("rechunk-p2p") for key in x2.__dask_keys__())
+
+    assert x2.chunks == new
+    assert np.all(await c.compute(x2) == a)
+
+
+@gen_cluster(client=True, config={"optimization.fuse.active": False})
+async def test_rechunk_2d(c, s, *ws):
```

Comment:
IIUC rechunking currently only works when this is disabled, correct?

Reply:
Yes, unfortunately.